File: frontend/src/types/invoice.ts
export type InvoiceStatus = 'draft' | 'issued' | 'partially_paid' | 'paid' | 'cancelled';

export interface LineItemCreate {
  description: string;
  quantity: string;
  unit_price: string;
}

export interface LineItemRead {
  id: string;
  description: string;
  quantity: string;
  unit_price: string;
  tax_rate_snapshot: string | null;
  line_subtotal: string;
  line_tax: string;
  line_total: string;
}

export interface InvoiceRead {
  id: string;
  invoice_number: string | null;
  customer_id: string;
  status: InvoiceStatus;
  issue_date: string | null;
  due_date: string;
  currency: string;
  subtotal: string;
  tax_amount: string;
  total_amount: string;
  amount_paid: string;
  customer_name_snapshot: string | null;
  customer_email_snapshot: string | null;
  customer_address_snapshot: string | null;
  cancelled_at: string | null;
  cancellation_reason: string | null;
  created_at: string;
  updated_at: string;
  is_overdue: boolean;
  line_items: LineItemRead[];
}

export interface InvoiceCreate {
  customer_id: string;
  due_date: string;
  line_items: LineItemCreate[];
}

export interface InvoiceUpdate {
  due_date?: string;
  line_items?: LineItemCreate[];
}

export interface InvoiceIssueRequest {
  note?: string;
}

export interface InvoiceCancelRequest {
  reason: string;
}

export interface InvoiceListFilters {
  status?: InvoiceStatus;
  customer_id?: string;
}

File: backend/app/schemas/invoice.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from app.models.invoice import InvoiceStatus


class LineItemCreate(BaseModel):
    description: str
    quantity: Decimal
    unit_price: Decimal


class LineItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    description: str
    quantity: Decimal
    unit_price: Decimal
    tax_rate_snapshot: Decimal | None
    line_subtotal: Decimal
    line_tax: Decimal
    line_total: Decimal


class InvoiceCreate(BaseModel):
    customer_id: uuid.UUID
    due_date: date
    line_items: list[LineItemCreate]


class InvoiceUpdate(BaseModel):
    due_date: date | None = None
    line_items: list[LineItemCreate] | None = None


class InvoiceIssueRequest(BaseModel):
    note: str | None = None


class InvoiceCancelRequest(BaseModel):
    reason: str


class InvoiceRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    invoice_number: str | None
    customer_id: uuid.UUID
    status: InvoiceStatus
    issue_date: date | None
    due_date: date
    currency: str
    subtotal: Decimal
    tax_amount: Decimal
    total_amount: Decimal
    amount_paid: Decimal
    customer_name_snapshot: str | None
    customer_email_snapshot: str | None
    customer_address_snapshot: str | None
    cancelled_at: datetime | None
    cancellation_reason: str | None
    created_at: datetime
    updated_at: datetime
    is_overdue: bool = False
    line_items: list[LineItemRead] = []

    @classmethod
    def from_orm_with_overdue(cls, invoice) -> "InvoiceRead":
        from app.services.invoices import is_overdue

        obj = cls.model_validate(invoice)
        obj.is_overdue = is_overdue(invoice)
        obj.line_items = [LineItemRead.model_validate(li) for li in invoice.line_items]
        return obj
